### PR TITLE
fix(nextjs-missing-metadata): stop flagging pages that inherit layout metadata

### PR DESCRIPTION
**Pull request by rayhanadev**
## Why

`nextjs-missing-metadata` flagged App Router pages as "missing metadata for search previews" even when they inherit `metadata` / `generateMetadata` from a co-located or ancestor `layout.*`. Next.js merges metadata down the segment chain (root layout → … → page), so a page covered by a parent layout's title/description already has search-preview metadata — the warning was a false positive (#775).

Before:

```tsx
// app/layout.tsx
export const metadata: Metadata = {
  title: { default: "Acme", template: "%s | Acme" },
  description: "...",
};

// app/page.tsx — flagged even though it inherits the layout's title/description
export default function Page() {
  return <main>Home</main>;
}
```

After:

```tsx
// app/page.tsx — no longer flagged (inherits metadata from app/layout.tsx)
export default function Page() {
  return <main>Home</main>;
}
```

## What changed

- Added `hasAncestorMetadataLayout` util that walks up the App Router tree (bounded by `CROSS_FILE_DIRECTORY_WALK_MAX_LEVELS`, stopping at `app/`) and returns true when a co-located or ancestor `layout.*` exports `metadata` / `generateMetadata`. Reuses the existing `doesModuleExportName` rather than re-parsing.
- `nextjs-missing-metadata` consults it after the cheap in-file check, so it stays quiet when an ancestor layout supplies metadata.
- Pages with no metadata anywhere in the segment chain are still flagged.
- Added cross-file tests: co-located layout, nested/ancestor layout, `generateMetadata`, `export { metadata }` specifier form, plus the two must-still-flag cases (layout without metadata; no layout at all).

Eval: RDE not run — this is a targeted false-positive fix scoped to one existing rule's cross-file inheritance, covered by unit + integration tests.

## Test plan

- `vp test run src/plugin/rules/nextjs/nextjs-missing-metadata.cross-file.test.ts` — 6 passed
- `vp test run src/plugin/rules/nextjs` — 26 passed
- `vp test run packages/react-doctor/tests/run-oxlint/nextjs.test.ts` (integration) — 34 passed
- `pnpm typecheck` (oxlint-plugin-react-doctor) — clean
- `vp lint` + `vp fmt --check` on touched files — clean

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Low Risk**
> Targeted lint-rule false-positive fix with shared refactor of existing cross-file layout walking; behavior for truly missing metadata is preserved and covered by new tests.
> 
> **Overview**
> Fixes a **false positive** in `nextjs-missing-metadata` (#775): App Router pages are no longer warned when they inherit `metadata` or `generateMetadata` from a co-located or ancestor `layout.*`, matching Next.js segment metadata merging.
> 
> The rule now treats **export specifiers** on the page (`export { metadata }`, `export { x as metadata }`) as satisfying the in-file check, then runs a **bounded walk** up the `app/` tree via new `hasAncestorMetadataLayout`, which reuses `doesModuleExportName` on candidate layouts. Pages with no metadata anywhere in the chain still get flagged.
> 
> Ancestor layout walking is **factored** into shared `hasAncestorLayoutMatching` (with updated stop logic for nested `app/` route segments and `.mts`/`.mjs` layouts); `hasAncestorSuspenseLayout` delegates to it. Shared constants `LAYOUT_FILE_NAMES` and `METADATA_EXPORT_NAMES` centralize layout filenames and export names. Cross-file tests cover inheritance, `generateMetadata`, specifier forms, and remaining positive cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 224df6060236ed88a0db2207e9ff7e9e97ff7a39. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/find-ancestor-metadata-layout.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import * as path from "node:path";
+import { CROSS_FILE_DIRECTORY_WALK_MAX_LEVELS } from "../constants/thresholds.js";
+import { doesModuleExportName } from "./does-module-export-name.js";
+
+const LAYOUT_FILE_NAMES = ["layout.tsx", "layout.jsx", "layout.ts", "layout.js"];
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/find-ancestor-metadata-layout.ts]: Fixed in fffe4345. Extracted the shared walk into `hasAncestorLayoutMatching` and moved `LAYOUT_FILE_NAMES` (and `METADATA_EXPORT_NAMES`) into `constants/nextjs.ts`; both the metadata and suspense ancestor-layout helpers now import them, so the constant and the loop are no longer duplicated.

[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/find-ancestor-metadata-layout.ts]: Fixed in fffe4345. The shared `LAYOUT_FILE_NAMES` now includes `layout.mts` and `layout.mjs`, matching `PAGE_OR_LAYOUT_FILE_PATTERN`. This also fixes the same latent gap in the suspense ancestor walk. Added a `layout.mts` regression test.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-missing-metadata.ts [hunk 2]
```diff
@@ -36,13 +37,16 @@ export const nextjsMissingMetadata = defineRule<Rule>({
         return false;
       });
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-missing-metadata.ts]: Fixed in fffe4345. The in-file check now also inspects `ExportNamedDeclaration.specifiers`, so `export { metadata }`, `export { x as metadata }`, and `export { metadata } from './meta'` are detected (keyed on the exported name, so `export { metadata as somethingElse }` is correctly still flagged). Added regression tests for all three.

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/has-ancestor-layout-matching.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  const normalizedPage = pageFilename.replaceAll("\\", "/");
+  let currentDirectory = path.dirname(normalizedPage);
+
+  for (let level = 0; level < CROSS_FILE_DIRECTORY_WALK_MAX_LEVELS; level++) {
+    for (const layoutFileName of LAYOUT_FILE_NAMES) {
+      const layoutPath = path.join(currentDirectory, layoutFileName);
+      if (layoutPath.replaceAll("\\", "/") === normalizedPage) continue;
+      if (matchesLayout(layoutPath)) return true;
+    }
+    if (path.basename(currentDirectory) === "app") break;
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/has-ancestor-layout-matching.ts]: Fixed in 224df606. The walk now stops at the outermost `app/` directory instead of the first one, so a page under a nested `app` segment (e.g. the `/app` route at `app/app/page.tsx`) reaches the real root `app/layout.tsx`. Added a regression test; this also fixes the same edge case in the Suspense ancestor walk.
